lib/plug/csrf_protection.ex

```diff
@@ -103,17 +109,27 @@ defmodule Plug.CSRFProtection do
 
   def init(opts), do: opts
 
-  def call(conn, _opts) do
+  def call(conn, opts) do
     csrf_token = get_session(conn, "_csrf_token")
     Process.put(:plug_csrf_token, csrf_token)
-
+    
     if not verified_request?(conn, csrf_token) do
-      raise InvalidCSRFTokenError
+      conn = case Keyword.get(opts, :with, :exception) do
+        :exception   -> raise InvalidCSRFTokenError
+        :nil_session -> clear_session_for_conn(conn)
+        _ -> raise ArgumentError, message: "Option :with should be one of :exception or :nil_session"
+      end
     end
 
     register_before_send(conn, &ensure_same_origin_and_csrf_token!(&1, csrf_token))
   end
 
+  #do not persist changes to this session
+  defp clear_session_for_conn(conn) do
+    conn = configure_session(conn, ignore: true)
```

Comment:
if we do that the attacker can, for eg, logout a legitimate session. There is a third strategy that we can implement which resets the session.

This is what rails does by default. 
https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/metal/request_forgery_protection.rb

and we should not be reading the reading the csrf token value from the session, but only from the request param or http header. The idea is to ensure that a request coming to the server from the attacking site (say by using an image tag) may contain the cookie but does not know the token.

if i'm not missing something doing `csrf_token = get_session(conn, "_csrf_token")` makes us vulnerable to CSRF attack